src/libraries/System.Private.CoreLib/src/System/DateOnly.cs

```diff
@@ -0,0 +1,846 @@
+        /// </summary>
+        /// <param name="value">The object to compare to this instance.</param>
+        /// <returns>true if the value parameter equals the value of this instance; otherwise, false.</returns>
+        public bool Equals(DateOnly value) => _dayNumber == value._dayNumber;
+
+        /// <summary>
+        /// Returns a value indicating whether this instance is equal to a specified object.
+        /// </summary>
+        /// <param name="value">The object to compare to this instance.</param>
+        /// <returns>true if value is an instance of DateOnly and equals the value of this instance; otherwise, false.</returns>
+        public override bool Equals(object? value) => value is DateOnly dateOnly && _dayNumber == dateOnly._dayNumber;
+
+        /// <summary>
+        /// Returns the hash code for this instance.
+        /// </summary>
+        /// <returns>A 32-bit signed integer hash code.</returns>
+        public override int GetHashCode() => _dayNumber;
+
+        private const ParseFlags ParseFlagsDateMask = ParseFlags.HaveHour | ParseFlags.HaveMinute | ParseFlags.HaveSecond | ParseFlags.HaveTime | ParseFlags.TimeZoneUsed |
+                                                      ParseFlags.TimeZoneUtc | ParseFlags.CaptureOffset | ParseFlags.UtcSortPattern;
+
+        /// <summary>
+        /// Converts a memory span that contains string representation of a date to its DateOnly equivalent by using culture-specific format information and a formatting style.
+        /// </summary>
+        /// <param name="s">The memory span that contains the string to parse.</param>
+        /// <param name="provider">An object that supplies culture-specific format information about s.</param>
+        /// <param name="style">A bitwise combination of enumeration values that indicates the permitted format of s. A typical value to specify is None.</param>
+        /// <returns>An object that is equivalent to the date contained in s, as specified by provider and styles.</returns>
+        public static DateOnly Parse(ReadOnlySpan<char> s, IFormatProvider? provider = default, DateTimeStyles style = DateTimeStyles.None)
+        {
+            ParseOperationResult result = TryParseInternal(s, provider, style, out DateOnly dateOnly);
+            if (result != ParseOperationResult.Success)
+            {
+                switch (result)
+                {
+                    case ParseOperationResult.WrongStyles: throw new ArgumentException(SR.Argument_InvalidDateStyles, nameof(style));
+                    case ParseOperationResult.ParseFailure: throw new FormatException(SR.Format(SR.Format_BadDateOnly, s.ToString()));
+                    default:
+                        Debug.Assert(result == ParseOperationResult.WrongParts);
+                        throw new FormatException(SR.Format(SR.Format_DateTimeOnlyContainsNoneDateParts, s.ToString(), nameof(DateOnly)));
+                }
```

Comment:
This block of exception handling code is duplicated in multiple places.  It should instead be a helper method, just as with DateTime.  That not only saves on duplication, it makes the call site more streamlined, with rarely used code separated out, makes it more likely to be inlineable, etc.